File: bipartitegraph.py
# timecomplexity: O(n)
# leetcode : accepted
# no doubts

# a graph is said to be bipartite if the two edges of a line segmet are in two separate sets.

# carrying out a simple DFS


class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        colour = [-1] * n  # a visited matrix

        for node in range(n):
            if colour[node] == -1:  # if the node is not visitd
                stack = []
                stack.append(node)  # append the unvisted node into stack
                colour[node] = 0  # and make the node in the colour list as zero
                while (len(stack) != 0):

                    item = stack.pop()  # pop the node

                    #                     print(item)

                    #                     print(colour)

                    for neighbour in graph[item]:  # for all its neighbours
                        print("neighbour", neighbour)
                        if colour[neighbour] == -1:  # if the neighbours are unvisited

                            stack.append(neighbour)  # add them to  stack
                            colour[neighbour] = colour[item] ^ 1  # and invert the colours. 0->1  or 1 ->0
                            print(colour)

                        elif colour[neighbour] == colour[item]:  # ifs already visited, then its not a biparitiie
                            return False  # retuning flase
        return True

File: BipartiteGraph.cs
//https://leetcode.com/submissions/detail/640578784/
//Time: O(V + E)
//Space: O(n) for storing colors

public class Solution {
    bool isBipartite;
    
    public bool IsBipartite(int[][] graph) {
        isBipartite = true;
        //0 - NoColor, 1 - Red, -1 Blue
        int[] colors = new int[graph.Length];
        
         for(int i = 0; i < graph.Length; i++) {
            if(colors[i] == 0) {
                colors[i] = 1;
                Dfs(graph, colors, i);
            }
        }
        
        return isBipartite;
    }
    
    private void Dfs(int[][] graph, int[] colors, int node) {

        if(!isBipartite) {
            return;
        }
        
        for(int i = 0; i < graph[node].Length; i++) {
             
            if(colors[graph[node][i]] == 0) {
               
               if(colors[node] == 1){
                   colors[graph[node][i]] = -1;
               } else {
                 colors[graph[node][i]] = 1;
               }
               Dfs(graph, colors, graph[node][i]);
                
            } else {
                if(colors[graph[node][i]] == colors[node]) {
                    isBipartite = false;
                    return;
                }                    
            } 
        }        
    }
    
}